pkg/controllers/node/health/controller.go

```diff
@@ -130,3 +161,57 @@ func (c *Controller) annotateTerminationGracePeriod(ctx context.Context, nodeCla
 
 	return nil
 }
+
+// isNodePoolHealthy checks if the number of unhealthy nodes managed by the given NodePool exceeds the health threshold.
+// defined by the cloud provider
+// Up to 20% of Nodes may be unhealthy before the NodePool becomes unhealthy (or the nearest whole number, rounding up).
+// For example, given a NodePool with three nodes, one may be unhealthy without rendering the NodePool unhealthy, even though that's 33% of the total nodes.
+// This is analogous to how minAvailable and maxUnavailable work for PodDisruptionBudgets: https://kubernetes.io/docs/tasks/run-application/configure-pdb/#rounding-logic-when-specifying-percentages.
+func (c *Controller) isNodePoolHealthy(ctx context.Context, nodePoolName string) (bool, error) {
+	nodeList := &corev1.NodeList{}
+	if err := c.kubeClient.List(ctx, nodeList, client.MatchingLabels(map[string]string{v1.NodePoolLabelKey: nodePoolName})); err != nil {
+		return false, err
+	}
+	nodePool := &v1.NodePool{}
+	if err := c.kubeClient.Get(ctx, types.NamespacedName{Name: nodePoolName}, nodePool); err != nil {
+		return false, err
+	}
+	stored := nodePool.DeepCopy()
+
+	healthy := c.isHealthyForNodes(nodeList.Items)
+	if !healthy {
+		nodePool.StatusConditions().SetTrueWithReason(v1.ConditionTypeHealthy, "Healthy", fmt.Sprintf("more then %s nodes are unhealthy", allowedUnhealthyPercent.String()))
+	} else {
+		_ = nodePool.StatusConditions().Clear(v1.ConditionTypeHealthy)
```

Comment:
I don't think that we should clear it -- the whole point of this condition is that we mark it as True when the NodePool is healthy -- further, we need to make sure that we mark the NodePool as healthy when there are no nodes, which is something that we are going to have to probably codify into another controller -- I actually wonder if we just hang this logic on another controller entirely (nodepool status controller) and just validate whether a NodePool is or isn't healthy when we are making a decision on these nodes

Reply:
Discussed offline, dropping the observability aspect from this PR and revising at a later time